lib/Conversion/ExportVerilog/ExportVerilog.cpp

```diff
@@ -507,30 +520,114 @@ getLocationInfoAsStringImpl(const SmallPtrSetImpl<Attribute> &locationSet) {
     }
     sstr << '}';
   }
-  return sstr.str();
 }
 
-static std::string
-getLocationInfoAsStringImpl(const SmallPtrSetImpl<Attribute> &locationSet,
-                            LoweringOptions::LocationInfoStyle style) {
+/// Return the location information as a (potentially empty) string.
+static void
+emitLocationSetInfoImpl(llvm::raw_string_ostream &sstr,
+                        const SmallPtrSetImpl<Attribute> &locationSet,
+                        LoweringOptions::LocationInfoStyle style) {
+  // Fast pass some common cases.
+  switch (locationSet.size()) {
+  case 1:
+    emitLocationInfo(sstr, cast<LocationAttr>(*locationSet.begin()), style);
+    [[fallthrough]];
+  case 0:
+    return;
+  default:
+    break;
+  }
+
+  // Sort the entries into distinct location printing kinds.
+  SmallVector<FileLineColLoc, 8> flcLocs;
+  SmallVector<Attribute, 8> otherLocs;
+  flcLocs.reserve(locationSet.size());
+  otherLocs.reserve(locationSet.size());
+  for (Attribute loc : locationSet) {
+    if (auto flcLoc = loc.dyn_cast<FileLineColLoc>())
+      flcLocs.push_back(flcLoc);
+    else
+      otherLocs.push_back(loc);
+  }
```

Comment:
I think the iteration order of `locationSet` is non-deterministic so `otherLocs` needs to be sorted (`flcLocs` is sorted in `printFileLineColSetInfo`). Or alternatively I think we can use `SetVector` for `locationSet`

Reply:
Good catch! Refactored such that both vectors are sorted by a generic sorting routine that considers all location types.